crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -2268,6 +2289,100 @@ where
+    fn future_annotations_or_stub(&self) -> bool {
+        self.has_future_annotations
+    }
+
+    fn python_version(&self) -> PythonVersion {
+        self.python_version
+    }
+
+    fn with_source<T>(&self, f: impl FnOnce(&str) -> T) -> T {
+        let src = source_text(self.db.upcast(), self.file);
+        f(src.as_str())
+    }
+
+    // TODO(brent) handle looking up `global` bindings
+    fn global(&self, _name: &str) -> Option<TextRange> {
+        None
+    }
+
+    fn in_async_context(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            match scope.kind() {
+                ScopeKind::Class | ScopeKind::Lambda => return false,
+                ScopeKind::Function => {
+                    return scope.node().expect_function().is_async;
+                }
+                ScopeKind::Comprehension
+                | ScopeKind::Module
+                | ScopeKind::TypeAlias
+                | ScopeKind::Annotation => {}
+            }
+        }
+        false
+    }
+
+    fn in_await_allowed_context(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            match scope.kind() {
+                ScopeKind::Class => return false,
+                ScopeKind::Function | ScopeKind::Lambda => return true,
+                ScopeKind::Comprehension
+                | ScopeKind::Module
+                | ScopeKind::TypeAlias
+                | ScopeKind::Annotation => {}
+            }
+        }
+        false
+    }
+
+    fn in_sync_comprehension(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            let generators = match scope.node() {
+                NodeWithScopeKind::ListComprehension(node) => &node.generators,
+                NodeWithScopeKind::SetComprehension(node) => &node.generators,
+                NodeWithScopeKind::DictComprehension(node) => &node.generators,
+                _ => continue,
+            };
+            if generators.iter().all(|gen| !gen.is_async) {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn in_module_scope(&self) -> bool {
+        self.scope_stack.len() == 1
+    }
+
+    fn in_function_scope(&self) -> bool {
+        let kind = self.scopes[self.current_scope()].kind();
+        matches!(kind, ScopeKind::Function | ScopeKind::Lambda)
+    }
+
+    fn in_generator_scope(&self) -> bool {
+        matches!(
+            self.scopes[self.current_scope()].node(),
+            NodeWithScopeKind::GeneratorExpression(_)
+        )
+    }
+
+    fn in_notebook(&self) -> bool {
+        self.file.source_type(self.db.upcast()).is_ipynb()
+    }
+
+    fn report_semantic_error(&self, error: SemanticSyntaxError) {
+        if self.db.is_file_open(self.file) {
```

Comment:
You could consider changing the `semantic_checker` field on the `SemanticIndexBuilder` to be an `Option<SemanticSyntaxChecker>` and only initialize it with `Some` if the file is open. `with_semantic_checker` is a no-op if the checker is `None`.